app/controllers/bz_controller.rb

```diff
@@ -673,18 +673,24 @@ def user_retained_data_batch
 
     data = {}
     if params[:names]
-      params[:names].each do |name|
-        next if data[name]
-        result = RetainedData.where(:user_id => user.id, :name => name)
-        if params[:include_timings]
+
+      # Add the fields that are set in the database
+      fields = RetainedData.where(:user_id => user.id, :name => params[:names]).pluck(:name, :value, :updated_at)
+      if params[:include_timings]
+        fields.each do |f| 
           obj = {}
-          obj["value"] = result.empty? ? '' : result.first.value
-          obj["timestamp"] = result.empty? ? 0 : result.first.updated_at.to_i
-          data[name] = obj
-        else
-          data[name] = result.empty? ? '' : result.first.value
+          obj["value"] = f[1]
+          obj["timestamp"] = f[2].to_i
+          data[f[0]] = obj
         end
+      else
+        fields.each { |f| data[f[0]] = f[1] }
       end
+
+      # Now add the fields not in the database, but need empty values.
+      empty_value = params[:include_timings] ? { "value" => '', "timestamp" => 0 } : ''
+      params[:names].each { |nm| data[nm] = empty_value unless data[nm] }
+
```

Comment:
Whoops, these changes were actually the stuff from this commit: [Optimize loading a page or assignment retained data (aka magic field).](https://github.com/beyond-z/canvas-lms/commit/63ca41805cd19d596248de6ce5259370d03d7d51)